pkg/store/proxy.go

```diff
@@ -94,70 +113,82 @@ func (s *ProxyStore) Series(r *storepb.SeriesRequest, srv storepb.Store_SeriesSe
-			storeID := fmt.Sprintf("%v", storepb.LabelsToString(st.Labels()))
-			if storeID == "" {
-				storeID = "Store Gateway"
+	g.Go(func() error {
+		var (
+			seriesSet      []storepb.SeriesSet
+			storeDebugMsgs []string
+			r              = &storepb.SeriesRequest{
+				MinTime:                 r.MinTime,
+				MaxTime:                 r.MaxTime,
+				Matchers:                newMatchers,
+				Aggregates:              r.Aggregates,
+				MaxResolutionWindow:     r.MaxResolutionWindow,
+				PartialResponseDisabled: r.PartialResponseDisabled,
 			}
-			err = errors.Wrapf(err, "fetch series for %s %s", storeID, st)
-			if r.PartialResponseDisabled {
-				level.Error(s.logger).Log("err", err, "msg", "partial response disabled; aborting request")
-				return err
+			wg = &sync.WaitGroup{}
+		)
+
+		defer func() {
+			wg.Wait()
+			closeFn()
+		}()
+
+		for _, st := range stores {
+			// We might be able to skip the store if its meta information indicates
+			// it cannot have series matching our query.
+			// NOTE: all matchers are validated in labelsMatches method so we explicitly ignore error.
+			if ok, _ := storeMatches(st, r.MinTime, r.MaxTime, r.Matchers...); !ok {
+				storeDebugMsgs = append(storeDebugMsgs, fmt.Sprintf("store %s filtered out", st))
+				continue
 			}
-			respCh <- storepb.NewWarnSeriesResponse(err)
-			continue
-		}
+			storeDebugMsgs = append(storeDebugMsgs, fmt.Sprintf("store %s queried", st))
 
-		seriesSet = append(seriesSet, startStreamSeriesSet(sc, respCh, 10))
-	}
+			sc, err := st.Series(gctx, r)
+			if err != nil {
+				storeID := fmt.Sprintf("%v", storepb.LabelsToString(st.Labels()))
+				if storeID == "" {
+					storeID = "Store Gateway"
+				}
+				err = errors.Wrapf(err, "fetch series for %s %s", storeID, st)
+				if r.PartialResponseDisabled {
+					level.Error(s.logger).Log("err", err, "msg", "partial response disabled; aborting request")
+					return err
+				}
+				respSender.send(storepb.NewWarnSeriesResponse(err))
+				continue
+			}
 
-	level.Debug(s.logger).Log("msg", strings.Join(storeDebugMsgs, ";"))
-	g.Go(func() error {
-		defer close(respCh)
+			// Schedule streamSeriesSet that translates gRPC streamed response into seriesSet (if series) or respCh if warnings.
+			seriesSet = append(seriesSet, startStreamSeriesSet(gctx, wg, sc, respSender, st.String(), !r.PartialResponseDisabled))
+		}
+
+		level.Debug(s.logger).Log("msg", strings.Join(storeDebugMsgs, ";"))
 
 		if len(seriesSet) == 0 {
 			// This is indicates that configured StoreAPIs are not the ones end user expects
 			err := errors.New("No store matched for this query")
 			level.Warn(s.logger).Log("err", err, "stores", strings.Join(storeDebugMsgs, ";"))
-			respCh <- storepb.NewWarnSeriesResponse(err)
+			respSender.send(storepb.NewWarnSeriesResponse(err))
 			return nil
 		}
 
 		mergedSet := storepb.MergeSeriesSets(seriesSet...)
 		for mergedSet.Next() {
 			var series storepb.Series
 			series.Labels, series.Chunks = mergedSet.At()
```

Comment:
yes :+1: This came from the TSDB I think. Worth to contribute such there.. or actually hosting this in Thanos is important as well I think since we use it heavily. Where to put it though? Maybe blog post?